File: javascript/models/review.js
const Review = ( function Review(){
  var counter = 0

  return class {
    constructor(text, customerId, restaurantId) {
      this.id = ++counter
      this.text = text
      this.customerId = customerId
      this.restaurantId = restaurantId
      store.reviews.push(this)
    }

    customer(){
      return store.customers.filter((customer) => {
        return customer.id===this.customerId
      })
    }

    restaurant(){
      return store.restaurants.filter((restaurant) => {
        return restaurant.id===this.restaurantId
      })
    }

    static all() {
      return store.reviews
    }
  }
}
())

File: sql-relations/app/models/review.rb
class Review
  include Databaseable::InstanceMethods
  extend Databaseable::ClassMethods

  ATTRIBUTES = {
    id: "INTEGER PRIMARY KEY",
    customer_id: "INTEGER",
    restaurant_id: "INTEGER"
  }

  attr_accessor(*self.public_attributes)
  attr_reader :id


  def customer
    sql = <<-SQL
      SELECT customers.* FROM customers
      INNER JOIN reviews ON reviews.customer_id = customers.id
      WHERE reviews.id = ?
      LIMIT 1
    SQL
    self.class.db.execute(sql, self.id).first
  end

  def restaurant
    sql = <<-SQL
      SELECT restaurants.* FROM restaurants
      INNER JOIN reviews ON reviews.restaurant_id = restaurants.id
      WHERE reviews.id = ?
      LIMIT 1
    SQL
    self.class.db.execute(sql, self.id).first
  end


end
